File: c01/main.ml
module Zadanie1Easy = struct
  let flatten1 = List.fold_left (fun acc xs -> acc @ xs) []
end

module Zadanie2 = struct 
  (* let count_num n = List.fold_left (fun acc x -> if x = n then acc + 1 else acc) 0 *)
  let count_num e lst =
    let rec aux acc lst =
      if lst = [] then acc
      else if e = (List.hd lst) then aux (acc + 1) (List.tl lst) 
      else aux acc (List.tl lst)
    in aux 0 lst
end;;

module Zadanie3 = struct 
  let replicate str n =
    let rec aux acc n =
      if n = 0 then acc else aux (str :: acc) (n - 1)
    in
    if n < 0 then failwith "n < 0"
    else
    aux [] n

  (* let replicate str n = List.init n (fun _ -> str) *)
end;;

module Zadanie4 = struct 
  let sqrList xs = List.map (fun x -> x * x) xs
end;;

module Zadanie5 = struct 
  let palindrome xs = xs = List.rev xs
end;;

module Zadanie6 = struct 
  let list_length = List.fold_left (fun acc _ -> acc + 1) 0
end;;
    
module Zadanie7 = struct 
  let logb2 = 
    let rec aux acc n = 
      if n < 2 then acc else aux (acc + 1) (n / 2)
    in aux 0

  let rec solve = function
    | 1 -> 1
    | n -> solve @@ logb2 n + solve n / 2
end;;

module Zadanie1 = struct 
  let flatten_two_lists xs ys =
    let rec aux xs ys =
      match xs with
      | [] -> ys
      | h :: t -> aux t (h :: ys)
    in
    aux (List.rev xs) ys

  (* 1 opcja *)
  let flatten1 xss =
    let rec aux lists acc =
      match lists with
      | [] -> acc
      | h :: t -> aux t (flatten_two_lists h acc)
    in
    aux (List.rev xss) []

  (* 2 opcja *)
  let flatten1 xss =
    List.fold_left (fun acc xs -> flatten_two_lists xs acc) [] (List.rev xss)
end;;

File: c01/main.scala
object Main {
  def main(args: Array[String]): Unit = {
    // val lista = List(List(1,2,3), List(5,6,3), List(2,253,6))
    // val out = zadanie1.flatten1(lista)
    // println(out)

    // val lista2 = List(1,2,3,3,4,3,3,5,3,2)
    // val out2 = zadanie2.countNum(lista2, 3)
    // println(out2)

    val lista5 = List('b', 'a', 'l', 'a', 'b')
    val out = zadanie5.palindrome(lista5)
    println(out)
  }
}

object zadanie1 {
  def flattenTwoLists(xs: List[Int], ys: List[Int]): List[Int] = {
    def aux(xs: List[Int], ys: List[Int]): List[Int] = {
      xs match {
        case Nil => ys
        case h :: t => aux(t, h :: ys)
      }
    }
    aux(xs.reverse, ys)
  }

  // 1 opcja:
  def flatten1(xss: List[List[Int]]): List[Int] = {
    def aux(lists: List[List[Int]], acc: List[Int]): List[Int] = {
      lists match {
        case Nil => acc
        case h :: Nil => aux(Nil, flattenTwoLists(h, acc))
        case h :: t => aux(t, flattenTwoLists(h, acc))
      }
    }
    aux(xss.reverse, List())
  }

  // 2 opcja:
  // def flatten1(xss: List[List[Int]]): List[Int] = {
  //   xss.fold(List())(flattenTwoLists)
  // }

}

object zadanie2 {
  def countNum(l: List[Int], num: Int): Int = {
    l.fold(0)((result, curr) => (if curr == num then result + 1 else result))

    // def aux(l: List[Int], acc: Int): Int = {
    //   l match {
    //     case h :: t if h == num => aux(t, acc + 1)
    //     case _ :: t => aux(t, acc)
    //     case Nil => acc
    //   }
    // }
    // aux(l, 0)
  }
}

object zadanie3 {
  def replicate(str: String, n: Int): List[String] = {
    def aux(n: Int, acc: List[String]): List[String] = {
      if n == 0 then acc
      else if n < 0 then throw new Exception("n must be positive")
      else aux(n-1, str :: acc)
    }
    aux(n, List())
  }
}

object zadanie4 {
}

object zadanie5 {
  def palindrome(l: List[Char]): Boolean = l == l.reverse
}

object zadanie6 {
  def listLength[A](xs: List[A]): Int = {
    xs match {
      case Nil => 0
      case _ :: t => 1 + listLength(t)
    }
  }
}

object zadanie7 {
  // dlugie
}

